enterprise/backend/src/metabase_enterprise/data_editing/coerce.clj

```diff
@@ -36,15 +48,64 @@
 (defn- json-zdt->time [s]
   (-> s ZonedDateTime/parse .toLocalTime str))
 
-(def input-coercion-fn
-  "Maps a coercion strategy to a function from input JSON object to the action representation (to be supplied to JDBC).
-
-  Assumes the input JSON object is valid for the coercion strategy, and can fit within the bounds of the target column."
-  {:Coercion/UNIXSeconds->DateTime          #'json-zdt->unix-seconds
-   :Coercion/UNIXMilliSeconds->DateTime     #'json-zdt->unix-millis
-   :Coercion/UNIXMicroSeconds->DateTime     #'json-zdt->unix-micros
-   :Coercion/UNIXNanoSeconds->DateTime      #'json-zdt->unix-nanos
-   :Coercion/YYYYMMDDHHMMSSString->Temporal #'json-zdt->yyyymmddhhmmss
-   :Coercion/ISO8601->DateTime              identity
-   :Coercion/ISO8601->Date                  #'json-zdt->date
-   :Coercion/ISO8601->Time                  #'json-zdt->time})
+(defn- millis->json-zdt
+  [millis]
+  (-> millis
+      java.time.Instant/ofEpochMilli
+      (java.time.ZonedDateTime/ofInstant (java.time.ZoneId/systemDefault))
+      str))
+
+(defn- unix-seconds->json-zdt
+  [unix-seconds]
+  (millis->json-zdt (* unix-seconds 1000)))
+
+(defn- unix-millis->json-zdt
+  [unix-millis]
+  (millis->json-zdt unix-millis))
+
+(defn- unix-micros->json-zdt
+  [unix-micros]
+  (millis->json-zdt (quot unix-micros 1000)))
+
+(defn- unix-nanos->json-zdt
+  [unix-nanos]
+  (millis->json-zdt (quot unix-nanos 1000000)))
+
+(defn- date->json-zdt
+  [date]
+  (-> date
+      java.time.LocalDate/parse
+      (.atStartOfDay (java.time.ZoneId/systemDefault))
+      str))
+
+(defn- time->json-zdt
+  [time]
+  (-> time
+      java.time.LocalTime/parse
+      (.atDate (java.time.LocalDate/now))
+      (.atZone (java.time.ZoneId/systemDefault))
+      str))
+
+(def coercion-fns
+  "Maps a coercion strategy to a map containing both input and output conversion functions.
+
+   :input  - Function to convert from input JSON to database format
+   :output - Function to convert from database format to JSON output
+
+   Assumes the input/output values are valid for the coercion strategy and can fit within bounds."
+  {:Coercion/UNIXSeconds->DateTime          {:in  #'json-zdt->unix-seconds
+                                             :out #'unix-seconds->json-zdt}
```

Comment:
I was really hoping we could reuse the logic in QP, but it seems they do all the coercion in the query?

I guess one way we could get reuse would be to run an in-memory query using something like this:

```clojure
(toucan2.core/query
 {:with   [[[:t {:columns (keys row}]
            {:values [(vals row]}]]
  :select (for [f fields] [(get-the-honey-fragment f) (:name f)])
  :from   [:t]})
```

Not sure it's worth being so obtuse just to reuse all the logic though 😢

Reply:
that will actually get the value "for the DB" which is the "in" direction right? we want "out" for this use case.